# SURF-2018

An effort from a little volunteer of SURF-2018 project. -- Aug 2018

## What is SURF?

The Summer Undergraduate Research Fellowships (SURF) programme offers undergraduate students the chance to work on research projects in collaboration with academic supervisors, during the summer holidays in XJTLU.
For more information, see [here](https://www.xjtlu.edu.cn/en/research/summer-undergraduate-research-fellowships).

## Research Title

Predict the target specificity of RNA methylation reader, writer and eraser (a machine learning/classification project)

## Research Abstract

RNA methylation has emerged as an important layer of gene regulation. Tens of thousands of RNA methylation sites are regulated by a small number of its regulators, including the readers (METTL3 and METTL16), erasers (FTO and ALKBH5) and readers (YTHDF1, YTHDF2 and YTHDF3). The proposed project seek to computational predict the substrate specificity of readers, writers and erasers so as to understand the regulatory mechanism of the RNA methylome. We will try to use different machine learning approaches to make the prediction and validate the results with cross-validation.

## My Effort:

-- **Learn basic concept of machine learning.**
There are some sources I refered to:
1. [Machine learning](https://en.wikipedia.org/wiki/Machine_learning)
2. [Cross validation](https://en.wikipedia.org/wiki/Cross-validation_(statistics))
3. [Gradient boosting](https://en.wikipedia.org/wiki/Gradient_boosting)
4. [Random forest](https://en.wikipedia.org/wiki/Random_forest)
5. [Support-vector machine](https://en.wikipedia.org/wiki/Support-vector_machine)

-- **Learn application of machine learning method in bioinformatics (focusing on epigenetics such as posttranscriptional modification of RNAs).**
I detailed read some papers that are closely related to this project:
1. [Dynamic transcriptomic m6A decoration writers, erasers, readers and functions in RNA metabolism](file:///D:/git/SURF-2018/paper/Dynamic%20transcriptomic%20m6A%20decoration%20writers,%20erasers,%20readers%20and%20functions%20in%20RNA%20metabolism.pdf)
2. [MethyRNA A web-server for identification of N6-methyladenosine sites](file:///D:/git/SURF-2018/paper/MethyRNA%20A%20web-server%20for%20identification%20of%20N6-methyladenosine%20sites.pdf)
3. [SRAMP prediction of mammalian N6 methyladenosine (m6A) sites based on sequence derived features](file:///D:/git/SURF-2018/paper/SRAMP%20prediction%20of%20mammalian%20N6%20methyladenosine%20(m6A)%20sites%20based%20on%20sequence%20derived%20features.pdf)

-- **Learn R programming, visualization and caret package.**
I read these two books recommended by Dr. Meng:
1. [An introduction to R](file:///D:/git/SURF-2018/R%20programming%20books/R-intro.pdf)


